sdks/python/apache_beam/runners/common.py

```diff
@@ -51,117 +51,150 @@ def receive(self, windowed_value):
-    self.main_receivers = as_receiver(tagged_receivers[None])
+  A DoFnInvoker describes a particular way for invoking methods of a DoFn
+  represented by a given DoFnSignature."""
 
-    # TODO(sourabh): Deprecate the use of context
-    if state:
-      assert context is None
-      self.context = DoFnContext(self.step_name, state=state)
-    else:
-      assert context is not None
-      self.context = context
+  def __init__(self, signature):
+    self.signature = signature
 
-    class ArgPlaceholder(object):
-      def __init__(self, placeholder):
-        self.placeholder = placeholder
+  @staticmethod
+  def create_invoker(
+      signature, use_simple_invoker, context, side_inputs, input_args,
+      input_kwargs):
+    if use_simple_invoker:
+      return SimpleInvoker(signature)
+    else:
+      return PerWindowInvoker(
+          signature, context, side_inputs, input_args, input_kwargs)
 
-    # Stash values for use in dofn_process.
-    self.side_inputs = side_inputs
-    self.has_windowed_inputs = not all(
-        si.is_globally_windowed() for si in self.side_inputs)
+  def invoke_process(self, element, process_output_fn):
+    raise NotImplementedError
 
-    self.args = args if args else []
-    self.kwargs = kwargs if kwargs else {}
-    self.dofn = fn
-    self.dofn_process = fn.process
+  def invoke_start_bundle(self, process_output_fn):
+    defaults = self.signature.start_bundle_method.defaults
+    defaults = defaults if defaults else []
+    args = [self.context if d == core.DoFn.ContextParam else d
+            for d in defaults]
+    process_output_fn(None, self.signature.start_bundle_method(*args))
 
-    arguments, _, _, defaults = self.dofn.get_function_arguments('process')
+  def invoke_finish_bundle(self, process_output_fn):
+    defaults = self.signature.start_bundle_method.defaults
     defaults = defaults if defaults else []
-    self_in_args = int(self.dofn.is_process_bounded())
+    args = [self.context if d == core.DoFn.ContextParam else d
+            for d in defaults]
+    process_output_fn(None, self.signature.finish_bundle_method(*args))
 
-    self.use_simple_invoker = (
-        not side_inputs and not args and not kwargs and not defaults)
-    if self.use_simple_invoker:
-      # As we're using the simple invoker we don't need to compute placeholders
-      return
 
+class SimpleInvoker(DoFnInvoker):
+  """An invoker that processes elements ignoring windowing information."""
+
+  def invoke_process(self, element, process_output_fn):
+    process_output_fn(element, self.signature.process_method(element.value))
+
+
+class PerWindowInvoker(DoFnInvoker):
+  """An invoker that processes elements considering windowing information."""
+
+  def __init__(self, signature, context, side_inputs, input_args, input_kwargs):
+    super(PerWindowInvoker, self).__init__(signature)
+    self.side_inputs = side_inputs
+    self.context = context
+    self.has_windowed_inputs = not all(
```

Comment:
I wonder if the has_windowed_inputs bit should instead be a third class, not a bit in the second Invoker class.